bigchaindb_driver/pool.py

```diff
@@ -46,8 +51,13 @@ def pick(self, connections):
                 :class:`~bigchaindb_driver.connection.Connection` instances.
 
         """
-        self.picked += 1
-        self.picked = self.picked % len(connections)
+
+        current_time_ms = datetime.utcnow()
+
+        while current_time_ms <= connections[self.picked]['time']:
```

Comment:
We should only check each connection once here to avoid running late with respect to the user-specified timeout.

Reply:
I agree on the fact that it is actually failing! (ill fix it, need to think a test on that too i guess) is not respecting the user-specified timeout, but disagree on checking only once, more specifically:

1. We need to add the timeout check on that point too, (not figure out how yet).

2. But we could not stop the loop after calling only next nodes related to the connections length nor all of them once, as we don't know which node gets ready after its specific backoff time, couple of examples:

- When all nodes for a little while stop responding, and we need to find the very first one asap.

- When we are on hitting the last node, say ```ith= len(nodes)```, suddenly that one goes down and the next node waking up is ```ith-1```, we need to start hitting that one.
 
what do you think?

Reply:
Good point!

I see 2 approaches:

1. Update the time counter on every iteration of the loop in `forward_request`.
```
end = time()
time_left -= timedelta(seconds=end - start)
```
It is then fine to only check the nodes list once since it does not take a significant amount of time. If nothing is found the search will be simply repeated.

2. Pass the timeout (`utcnow() + time_left` in `forward_request`) to `pick` so that the while loop turns into `while current_time_ms <= connections[self.picked]['time'] and current_time_ms <= timeout`. Then you also need to alter `time_left` after the connection is found but before you make an HTTP request.